src/core/listener.c

```diff
@@ -643,6 +644,55 @@ QuicListenerClaimConnection(
         return FALSE;
     }
 
+    const uint8_t* NewNegotiatedAlpn = Event.NEW_CONNECTION.NewNegotiatedAlpn;
+    if (NewNegotiatedAlpn) {
+        const uint8_t* ListStart = Info->ClientAlpnList;
+        const uint8_t* ListEnd = Info->ClientAlpnList + Info->ClientAlpnListLength;
+        if (ListStart > NewNegotiatedAlpn ||
+            ListEnd <= NewNegotiatedAlpn ||
+            NewNegotiatedAlpn + NewNegotiatedAlpn[0] + sizeof(uint8_t) > ListEnd
+        ) {
+            QuicTraceEvent(
+                ListenerError,
+                "[list][%p] ERROR, %s.",
+                Listener,
+                "'NewNegotiatedAlpn' field is out of bounds of the 'ClientAlpnList' buffer.");
+            QuicConnTransportError(
+                Connection,
+                QUIC_ERROR_INTERNAL_ERROR);
+            return FALSE;
+        }
+
+        //
+        // Free current ALPN buffer if it's allocated on heap.
+        //
+        if (Connection->Crypto.TlsState.NegotiatedAlpn != Connection->Crypto.TlsState.SmallAlpnBuffer) {
+            CXPLAT_FREE(Connection->Crypto.TlsState.NegotiatedAlpn, QUIC_POOL_ALPN);
```

Comment:
You need to NULL after free, in case you fail to alloc below, you'd never clear the pointer.

Reply:
I'll do it after the pipeline.